Add std::tuple utility

nilsvu opened the pull request:
## Proposed changes

Utility functions for std::tuple metaprogramming

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe commented:
Would rvalue overloads for the slicing functions be useful?

nilsdeppe commented on tests/Unit/Utilities/Test_Tuple.cpp:
```diff
@@ -21,28 +21,55 @@ struct tuple_fold_plus {
   }
```
Can you please separate out formatting this file from the changes you're making? This is rather difficult to parse

nilsvu replied:
sure, done

nilsdeppe commented on tests/Unit/Utilities/Test_Tuple.cpp:
```diff
@@ -21,9 +21,35 @@ struct tuple_fold_plus {
 }  // namespace
 
-SPECTRE_TEST_CASE("Unit.Utilities.tuple_fold", "[Utilities][Unit]") {
+SPECTRE_TEST_CASE("Unit.Utilities.Tuple", "[Utilities][Unit]") {
```
Instead of just indenting these and having one giant function, why not have separate functions in the anonymous namespace that are then called from a since test? This is the pattern we have adopted in most places

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,33 @@ constexpr inline void tuple_transform(
       std::make_index_sequence<sizeof...(Elements)>{}, args...);
 }
+
+namespace tuple_impl_detail {
```
Other than the fact they these operate on tuples, they seem to have nothing to do with the other functions in these files. How about a `TupleSlice` file?

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,33 @@ constexpr inline void tuple_transform(
+
+}  // namespace tuple_impl_detail
+
+/// The subset of elements in `tuple` from index `Start` to (excluding) `Stop`
```
These should be in the `UtilitiesGroup` otherwise I doubt anyone can find them in the documentation

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,33 @@ constexpr inline void tuple_transform(
+
+/// The subset of elements in `tuple` from index `Start` to (excluding) `Stop`
+template <size_t Start, size_t Stop, typename... AllElements>
+auto tuple_slice(const std::tuple<AllElements...>& tuple) noexcept {
```
you need an rvalue overload

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -204,3 +204,19 @@ auto tuple_tail(const std::tuple<AllElements...>& tuple) noexcept {
+}  // namespace tuple_impl_detail
+
+template <typename Elements>
+using tuple_from_typelist =
```
Please document this and stick it in its own header file

nilsdeppe commented on tests/Unit/Utilities/Test_Tuple.cpp:
```diff
@@ -262,4 +263,9 @@ SPECTRE_TEST_CASE("Unit.Utilities.Tuple", "[Utilities][Unit]") {
   }
+  {
+    INFO("From typelist");
+    static_assert(std::is_same_v<tuple_from_typelist<tmpl::list<int, float>>,
```
check where the same type occurs multiple times, you have only 1 element, and you have 0 elements

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -204,3 +204,19 @@ auto tuple_tail(const std::tuple<AllElements...>& tuple) noexcept {
 }
+
+namespace tuple_impl_detail {
+
```
do we need the blank lines at the start and end of the namespaces?

nilsvu replied:
I find that more readable 🤷‍♂️ helps to see where namespaces start and end, and where classes and functions start, imo

nilsvu commented:
@nilsdeppe @wthrowe pushed fixups, also adding rvalue overloads. not sure if those are quite correct, please take a look. Thanks!

nilsdeppe commented on src/Utilities/TupleSlice.hpp:
```diff
@@ -28,17 +35,41 @@ auto tuple_slice(const std::tuple<AllElements...>& tuple) noexcept {
+template <size_t Start, size_t Stop, typename... AllElements>
+auto tuple_slice(std::tuple<AllElements...>&& tuple) noexcept {
+  return tuple_impl_detail::slice_impl<Start>(
+      std::forward<std::tuple<AllElements...>>(tuple),
```
you can move the tuple here and below too since you know you own it. `std::forward` is if you have forwarding references, not rvalue references

nilsdeppe commented on tests/Unit/Utilities/Test_TupleSlice.cpp:
```diff
@@ -9,8 +9,21 @@
-  CHECK(tuple_head<2>(tuple) == std::tuple<int, float>{1, 2.});
-  CHECK(tuple_tail<2>(tuple) == std::tuple<double, std::string>{3., ""});
+  {
+    std::tuple<int, float, double, std::string> tuple{1, 2., 3., ""};
```
Sorry for not thinking of this during my last review: can you test 1 and 0 element tuples? I'm not sure what they will do, so we should check that. Can you also test having `Size` in `head` and `tail` be larger than the size of the tuple? I think you'll find undesirable behavior in those cases and may want to add a `static_assert` :)

nilsvu replied:
Added the tests. The invalid `Size` and slice ranges trigger standard errors like this:

```
error: static_assert failed due to requirement '0UL < sizeof...(_Types)' "tuple_element index out of range"
```

I suppose that's as good as it gets and we don't need to add extra asserts?

nilsdeppe replied:
Are you sure that error isn't STL implementation dependent?

nilsvu replied:
no I'm not sure, can I check that? Would you suggest to add another assert then?

nilsvu replied:
with AppleClang it triggers here: `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:283:5`

nilsdeppe replied:
test in the container with libc++ and stdlibc++, and check what the standard says (cppreference should be good enough)

nilsvu replied:
[`std::get`](https://en.cppreference.com/w/cpp/utility/tuple/get) "Extracts the Ith element from the tuple. I must be an integer value in [0, sizeof...(Types))."

nilsdeppe replied:
Just add static asserts... Quoting cppreference doesn't help me, I can do that myself. My main concern is if this is used in generic code you'll end up creating a giant index sequence if something goes wrong, which just segfaults the compiler. I think you want to require the tuple have less than 1000 elements, and that the bounds are check

nilsvu replied:
Sure I‘ll add asserts, I just wasn’t sure before if that would be useful or add unnecessary code. In general, are static asserts cheap enough on compile time to add them liberally?

nilsdeppe commented on tests/Unit/Utilities/Test_TupleSlice.cpp:
```diff
@@ -0,0 +1,38 @@
+    CHECK(tuple_slice<1, 3>(std::move(tuple)) ==
+          std::tuple<float, double>{2., 3.});
+  }
+  {
```
In general your tests are much too simplistic. You should be testing tuples of rvalue references and lvalue references to make sure that works correctly and doesn't cause dangling references...

nilsvu commented on tests/Unit/Utilities/Test_TupleSlice.cpp:
```diff
@@ -35,4 +35,14 @@ SPECTRE_TEST_CASE("Unit.Utilities.TupleSlice", "[Utilities][Unit]") {
+    std::tuple<int&, float&&> tuple{b, std::move(c)};
+    auto sliced_tuple = tuple_slice<0, 2>(std::move(tuple));
+    CHECK(std::get<0>(sliced_tuple) == b);
+    CHECK(std::get<1>(sliced_tuple) == 2.);
```
This test is not yet correct of course. How can I test the rvalue reference is moved correctly? Do you know of an example where an rvalue overload is tested in the codebase?

nilsvu commented:
@nilsdeppe pushed fixups, could you take a look? In particular, could you check the rvalue overload test?

nilsdeppe commented on tests/Unit/Utilities/Test_TupleSlice.cpp:
```diff
@@ -35,4 +55,15 @@ SPECTRE_TEST_CASE("Unit.Utilities.TupleSlice", "[Utilities][Unit]") {
+    std::tuple<MoveTest&, MoveTest&&> tuple{b, MoveTest{}};
+    CHECK(std::get<0>(tuple).status == "initial");
+    CHECK(std::get<1>(tuple).status == "initial");
+    auto sliced_tuple = tuple_slice<0, 2>(std::move(tuple));
```
Where do you check that `b` was not moved?

nilsdeppe commented on tests/Unit/Utilities/Test_TupleSlice.cpp:
```diff
@@ -8,6 +8,26 @@
+  MoveTest() = default;
+  MoveTest(const MoveTest& /*rhs*/) noexcept { status = "copy-constructed"; }
+  MoveTest& operator=(const MoveTest& /*rhs*/) noexcept {
+    status = "copied";
```
You want to change the status in `rhs` since that's the object that is no longer "valid" after move.

nilsdeppe replied:
You are no longer using this status. Maybe add a comment why it's needed?

nilsvu commented:
@nilsdeppe @wthrowe I rewrote the slicing implementation with perfect forwarding in mind. Since the diff changed significantly I squashed the change. Could you take a look at the new implementation?

wthrowe commented on src/Utilities/TupleSlice.hpp:
```diff
@@ -0,0 +1,55 @@
+
+/// \ingroup UtilitiesGroup
+/// References to the subset of elements in `tuple` from index `Start` to
+/// (excluding) `Stop`
```
Either give an explicit return type or make the documentation more explicit about the result.  I assumed `tuple_slice<0, 1>(std::declval<const std::tuple<int, double>&>())` would have type `std::tuple<int>`, but this mentions references.

nilsvu replied:
The idea here is taken from https://ngathanasiou.wordpress.com/2016/11/28/want-a-std-slice/. The `tuple_slice` references lvalues and move-constructs rvalues. This also allows assigning to a slice, for instance, as well als forwarding a subset of arguments to a function: https://gist.github.com/picanumber/bba2339c5648b53355487cd97d2b70f7#file-slice_examples-cpp
The implementation with `std::forward_as_tuple` is very straight-forward. I'd be interested to hear yours and @nilsdeppe opinion on this!

wthrowe replied:
Ah, OK, I see.  This needs a lot more documentation, then.  That's not what I thought it was trying to do.  I'm also still not sure that I know what the return type will be if I pass it a temporary.

wthrowe reviewed (changes requested):
I'm going to hold off on reviewing the rest until I'm confident that I understand the intent of the code.

wthrowe commented on src/Utilities/TupleSlice.hpp:
```diff
@@ -0,0 +1,55 @@
+template <size_t Start, size_t Stop, typename Tuple>
+constexpr auto tuple_slice(Tuple&& tuple) noexcept {
+  constexpr size_t tuple_size = std::tuple_size<std::decay_t<Tuple>>::value;
+  static_assert(0 <= Start and Start <= Stop and Stop <= tuple_size,
```
`0 <= Start` is trivial.  I think some compilers warn about that sort of thing.

wthrowe commented on src/Utilities/TupleFromTypelist.hpp:
```diff
@@ -0,0 +1,26 @@
+/// \ingroup UtilitiesGroup
+/// A `std::tuple` from the types in a `tmpl::list`
+template <typename Elements>
+using tuple_from_typelist =
```
This is `tmpl::wrap<Elements, std::tuple>`.

nilsvu replied:
Since the implementation is so simple, do you think it's worth adding the alias at all?

wthrowe replied:
I vote no, but since tagged_tuple_from_typelist is already in the code I'll see if anyone else wants to weigh in.  @kidder @nilsdeppe anything to add?

nilsdeppe replied:
I guess one (big) reason I can see having the type alias is that, even though I've used Brigand a lot, I was unaware of `tmpl::wrap`. If Brigand actually had documentation I would definitely agree that we don't need such a type alias. My worry is that we will end up getting several implementations of `tuple_from_typelist` wherever it's needed locally if we don't have some way of documenting that `tmpl::wrap<Elements, std::tuple>` exists. I can certainly see myself (before being involved in this conversation) having added such a helper struct somewhere. Just something I think we need to consider when deciding here :)

nilsvu replied:
Perhaps we could contribute documentation to Brigand instead of adding aliases to our code.. could be the same amount of time spent in a more useful way. I have been quite annoyed by Brigand’s lack of documentation before, and I haven’t seen its documentation improve at all.

wthrowe replied:
You're right.  This keeps coming up, and we should just fix it.  I'll get some sort of cheat sheet together for the dev guide of the stuff we've found useful in SpECTRE.  That should solve our immediate problems.  We can work on upstreaming after that.

wthrowe replied:
Turns out there's a `tmpl::as_tuple` that does exactly this.

wthrowe reviewed (changes requested):
CI is giving a very scary-looking error that I assume is related to some of the test cases.  I'll look at them in more detail once that is resolved.

nilsvu committed:
Add tuple slicing utilities

nilsvu commented:
@￼nilsdeppe￼ ￼@wthrowe￼ Updated the implementation to this now:

```cpp
template <size_t Start, typename Tuple, size_t... Is>
constexpr auto slice_impl(Tuple&& tuple,
                          std::index_sequence<Is...> /*meta*/) noexcept {
  return std::tuple<std::tuple_element_t<Is + Start, std::decay_t<Tuple>>...>{
      std::get<Is + Start>(std::forward<Tuple>(tuple))...};
}
```

With this implementation `tuple_slice` behaves the same as `std::tuple_cat` for both lvalue tuples and rvalue tuples of lvalues, refs and rvalue refs. Thanks ￼@nilsdeppe￼ for suggesting to compare to `std::tuple_cat`!
I squashed because a lot of the diff changed, hope that's ok with you. Please take a look when you get the chance. (I’m not quite sure what caused the CI builds to die)

nilsvu commented:
@nilsdeppe @wthrowe could you take another look at this PR? I'm fine with either dropping or keeping the `tuple_from_typelist` commit, just let me know

wthrowe commented:
You still have a dangling reference in the test.

`tuple_from_typelist` is a duplicate of `tmpl::as_tuple`.  There's no reason to have our own copy.

nilsvu commented:
@wthrowe from which of the CI builds do you get that information? The CI logs appear in an interrupted state to me

Edit: Ran the test through ASAN and found the issue, pushed a fixup commit. Thanks @wthrowe

nilsdeppe commented on src/Utilities/TupleSlice.hpp:
```diff
@@ -0,0 +1,54 @@
+  return std::tuple<std::tuple_element_t<Is + Start, std::decay_t<Tuple>>...>{
+      std::get<Is + Start>(std::forward<Tuple>(tuple))...};
+}
+
```
why the blank line here and right after the namespace? That's 25% of this code chunk going to blank lines

nilsvu replied:
The blank lines make the code more readable I think 🤷‍♂️ Sometimes when reading through code I miss where namespaces begin and end, and the blank lines help to visually indicate that

nilsdeppe replied:
I thought that's what the `namespace` keyword and ending comment are for :D I think in this case it's a bit excessive since 25% of the lines are blank...

nilsvu replied:
fine :)

nilsdeppe commented on src/Utilities/TupleSlice.hpp:
```diff
@@ -0,0 +1,54 @@
+namespace tuple_impl_detail {
+
+template <size_t Start, typename Tuple, size_t... Is>
+constexpr auto slice_impl(Tuple&& tuple,
```
why not use trailing return type syntax instead of `auto`?

nilsvu replied:
Do you think d029f28 is better (and why)?

nilsdeppe replied:
Yes. I find code easier to read if it has the return type explicit. I generally dislike auto in return types because the rules are stupidly complicated for when it gives references, pointers, const, etc. I think being explicit whenever it is easy to do so is better :)

nilsdeppe commented on tests/Unit/Utilities/Test_TupleSlice.cpp:
```diff
@@ -0,0 +1,91 @@
+    ConstructionObserver a{};
+    ConstructionObserver& b = a;
+    ConstructionObserver c{};
+    std::tuple<ConstructionObserver, ConstructionObserver&,
```
what about `const &` and `const &&`? Should those work and be tested?

nilsvu replied:
added in f90bd0a

nilsdeppe reviewed (commented):
I'm happy for you to squash things

nilsdeppe reviewed (commented):
LGTM! Thanks for doing this! Please go ahead and squash :)

nilsvu commented:
@wthrowe I kept the fixup for the dangling reference issue around for you. Could you take another look at this PR? Thanks!

wthrowe commented:
Looks good.

nilsvu commented:
Rebased and squashed. Thanks for reviewing @nilsdeppe @wthrowe!
